### PR TITLE
every user should have an external_id

### PR DESCRIPTION
**Pull request by grosser**
@zendesk/samson ... unless anyone know why a user should not have an external_id ...

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -23,6 +23,7 @@ class User < ActiveRecord::Base
 
   before_create :set_token
   validates :time_format, inclusion: { in: TIME_FORMATS }
+  validates :external_id, presence: :true, unless: :integration?
```
Comment (jonmoter) [inline app/models/user.rb]: How about we validate on `create` only, so we don't end up with a situation where existing users become unable to save their user objects?

Comment (grosser) [inline app/models/user.rb]: users without external_id cannot log in ... so they are bogus records that should be invalid ... and if they somehow can log in I'd like to understand why / have them report an issue

Comment (jonmoter) [inline app/models/user.rb]: If we add a validation like this, we should have a plan on what to do with users that fail the validation.  Not okay to have a bunch of saved records in the database that aren't considered valid anymore.

**Comment by grosser** [comment]
updated to add migration that will mark invalid users as deleted ... will affect 30 users

**Comment by grosser** [comment]
@jonmoter looks good ?

**Comment by grosser** [comment]
@jonmoter review ?

File: db/migrate/20161103221846_remove_invalid_users.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+class RemoveInvalidUsers < ActiveRecord::Migration[5.0]
+  class User < ActiveRecord::Base
+  end
+
+  def change
+    User.where(deleted_at: nil, external_id: nil, integration: false).update_all(deleted_at: Time.at(1478211603))
```
Comment (jonmoter) [inline db/migrate/20161103221846_remove_invalid_users.rb]: What the heck is that timestamp?

Comment (grosser) [inline db/migrate/20161103221846_remove_invalid_users.rb]: just picking a random timestamp so I have something unique to query in case things go wrong :D

Comment (grosser) [inline db/migrate/20161103221846_remove_invalid_users.rb]: 2016-11-03 15:20:03 -0700

**Pull request merged**